Content.Server/Ninja/Systems/BatteryDrainerSystem.cs

```diff
@@ -0,0 +1,115 @@
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<BatteryDrainerComponent, BeforeRangedInteractEvent>(OnInteract);
+    }
+
+    /// <summary>
+    /// Start do after for draining a power source.
+    /// Can't predict PNBC existing so only done on server.
+    /// </summary>
+    private void OnInteract(EntityUid uid, BatteryDrainerComponent comp, BeforeRangedInteractEvent args)
+    {
+        var target = args.Target;
+        if (comp.BatteryUid == null || !HasComp<PowerNetworkBatteryComponent>(target))
+            return;
+
+        // nicer for spam-clicking to not open apc ui, and when draining starts, so cancel the ui action
+        //args.Cancel();
+
+        if (IsBatteryFull(comp.BatteryUid.Value))
+        {
+            _popup.PopupEntity(Loc.GetString("battery-drainer-full"), uid, uid, PopupType.Medium);
+            return;
+        }
+
+        var doAfterArgs = new DoAfterArgs(uid, comp.DrainTime, new DrainDoAfterEvent(), target: target, eventTarget: uid)
+        {
+            BreakOnUserMove = true,
+            MovementThreshold = 0.5f,
+            CancelDuplicate = false,
+            AttemptFrequency = AttemptFrequency.StartAndEnd
+        };
+
+        _doAfter.TryStartDoAfter(doAfterArgs);
+    }
+
+    /// <inheritdoc/>
+    protected override void OnDoAfterAttempt(EntityUid uid, BatteryDrainerComponent comp, DoAfterAttemptEvent<DrainDoAfterEvent> args)
+    {
+        base.OnDoAfterAttempt(uid, comp, args);
+
+        if (comp.BatteryUid == null || IsBatteryFull(comp.BatteryUid.Value))
+            args.Cancel();
+    }
+
+    /// <inheritdoc/>
+    protected override bool TryDrainPower(EntityUid uid, BatteryDrainerComponent comp, EntityUid target)
+    {
+        if (comp.BatteryUid == null || !TryComp<BatteryComponent>(comp.BatteryUid.Value, out var battery))
+            return false;
+
+        if (!TryComp<BatteryComponent>(target, out var targetBattery) || !TryComp<PowerNetworkBatteryComponent>(target, out var pnb))
+            return false;
+
+        if (MathHelper.CloseToPercent(targetBattery.CurrentCharge, 0))
+        {
+            _popup.PopupEntity(Loc.GetString("battery-drainer-empty", ("battery", target)), uid, uid, PopupType.Medium);
+            return false;
+        }
+
+        var available = targetBattery.CurrentCharge;
+        var required = targetBattery.MaxCharge - targetBattery.CurrentCharge;
+        // higher tier storages can charge more
+        var maxDrained = pnb.MaxSupply * comp.DrainTime;
+        var input = Math.Min(Math.Min(available, required / comp.DrainEfficiency), maxDrained);
+        if (_battery.TryUseCharge(target, input, targetBattery))
+        {
+            var output = input * comp.DrainEfficiency;
+            _battery.SetCharge(comp.BatteryUid.Value, battery.CurrentCharge + output, battery);
+            Spawn("EffectSparks", Transform(target).Coordinates);
```

Comment:
Effects should be done on the client rather than spawning server entities for it, just use a message or something.

Reply:
is there a message for spawning clientside entities already to reuse